contents/java/JUnit.md

```diff
@@ -0,0 +1,226 @@
+
+```java
+public class Repeater {
+  // Returns copies of reference string separated by the '\n' character.
+  public static String repeat(String reference, int copies) {
+    String result = "";
+    for (int i = 0; i <= copies; i++) {
+      result += reference + "\n";
+    }
+    return result;
+  }
+}
+```
+
+Let's write a test to check the correctness of our code. 
+
+```java
+public class RepeaterTest {
+  @Test
+  public void repeatTest() {
+    assertEquals(repeat("a", 5), "a\na\na\na\na\n");
+  }
+}
+```
+<tip-box>:bulb: By convention, we append `Test` to the name of the tested class.</tip-box>
+
+That's it! We do not have to write any additional runners or helpers to take advantage of the powerful testing features. The `@Test` tag tells JUnit that the repeatTest method is meant to be a test. JUnit will then automatically execute our test(s) in sequence.
+
+Let's have a look at the test results. 
+
+<img src="{{baseUrl}}/contents/java/assertFailed.PNG" />
+Hmm, the test seems to fail when we run it.
+Did you notice the subtle bug in Repeater?
+
+
+We "accidentally" caused the loop to have an extra iteration! Without the use of unit tests, we would have to manually confirm that the output indeed has 5 lines, an action that is tedious to repeat and prone to human errors.
+
+Now, let's fix the code.
+
+```java
+public class Repeater {
+  // Returns copies of reference string separated by the '\n' character.
+  public static String repeat(String reference, int copies) {
+    String result = "";
+    for (int i = 0; i < copies; i++) { // Strictly less than copies!
+      result += reference + "\n";
+    }
+    return result;
+  }
+}
+```
+
+<img src="{{baseUrl}}/contents/java/assertPass.PNG" />
+Running the same unit test again, we now have a passing test and we can be <popover effect="fade" content="We've only done one simple test. While this test works, we cannot be assured that all other cases would work too. How will our code behave if a negative number is passed in?">somewhat assured</popover> that our repeat function works as intended.
+
+<p></p> 
+
+## Benefits of JUnit
+
+_____
+
+Here are some key reasons why developers should consider using JUnit:
+
+### Benefit 1: JUnit is well-integrated with development tools
+
+JUnit is supported by many popular IDEs for developing in Java:
+* IntelliJ IDEA
+* Eclipse
+* NetBeans
+* Visual Studio Code
+
+This makes it easy for developers to write and run tests.
+
+### Benefit 2: JUnit is well-established
+
+As compared to other testing tools for Java, JUnit
+* has a longer history
+* has a larger community (As of 2019, JUnit has more questions on StackOverflow as compared to other tools)
+* is more popular among developers (according to [Synk's 2018 report](https://snyk.io/blog/jvm-ecosystem-report-2018-tools/))
+
+This means that it is easier to find answers or get help with JUnit.
```

Comment:
As these two benefits don't have a lot of content under them, you can present them as two bullet points instead.